pkg/apis/hive/v1/agent/platform.go

```diff
@@ -0,0 +1,136 @@
+package agent
+
+import (
+	//corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Platform stores cluster configuration for an agent based install where hosts are booted and registered, ready
+// to be assigned and used in a cluster.
+type Platform struct {
+	BareMetal *BareMetal `json:"bareMetal"` // mandatory for now, possibly optional in future
+
+	// Networking is the configuration for the pod network provider in
+	// the cluster.
+	Networking *Networking `json:"networking,omitempty"`
+
+	// Proxy defines the proxy settings for the cluster.
+	// If unset, the cluster will not be configured to use a proxy.
+	// +optional
+	Proxy *Proxy `json:"proxy,omitempty"`
```

Comment:
since these fields are mostly from install config? Can we look at accepting an install-config with predetermined platform like `None` ?
Since these fields seem to be used for openshift-installer to generate the right ignition configs, we can use maybe use install config as input and require that the platform None be used in it.
assisted installer can use this install config + fields in cluster deployment to generate a valid install-config for generating ignition configs. Since we have to generate an install config and also allow users to provide some overrides, this could be used to pass on Networking and Proxy configuration which seems like is optional for most cases??

Reply:
+1

There doesn't appear to be anything specific to agent-based for these fields, and ClusterDeployment already has installConfigSecretRef - that should be the starting point for addressing this use case

Reply:
Notes from call: Assisted service today does not want to require users to have to provide an install config, this allows for better parameter validation etc. It does support (arbitrary) overrides. 

Install config and ignition config today are generated only when a user confirms to proceed with actual install.

InstallEnvironment could have an install config template as a starting point. Users can either retrieve this, add customizations, and provide directly to hive exactly as it works today. Hive does not know about the existence of InstallEnvironment. If none is provided, we generate one for you.

Today, virtually no one uses the install config and ignition config overrides and customizations, they just use the generated one assisted handles for them. Group on the call proposes that we drop this for now, it will not be possible to customize or provide your own install/ignition config directly with the Kube API, and we can add later if needed.

This means that configuration we have requirements for must get modeled directly in a Kube API. Proxy is not cluster specific and should fit well in the proposed InstallEnvironment CRD.

Reply:
It's pretty easy to imagine at some point that Hive will need to have an answer for "what CRD should I use to configure my company's proxy so that it is used for all installs run by Hive?"

But InstallEnvironment wouldn't be that API? It's an API purely for configuring the generated discovery ISO?

(Maybe I have missed where the InstallEnvironment idea is written down somewhere? TIA)

Reply:
I don't think it's documented yet just an idea we've been discussing a lot. We should probably change that.

Loosely it sounded to me like it was "what is required to generate an ISO". Today it would be cluster specific but hopefully not within a few releases. Things like proxy, MAC address to static IP mappings, etc. Others can probably cover it better than I, but we should sketch something out soon and decide where it would live.

Reply:
Ok, well it sounds to me like Hive inevitably should have an API for stuff like proxy and NTP that goes beyond the "what is required to generate an ISO" semantics

Reply:
I would say it can go beyond "what is required to generate an ISO".  It should contain properties of the environment as opposed to properties for a specific cluster installation.
Rule of thumb - I should be able to create different clusters from the InstallEnvironment.  So between installations, or if we have multiple installations from the same InstallEnvironment, we would have common settings for proxy, NTP, static IPs, etc.  We would have different parameters for the cluster's networking, for example, so that goes into the ClusterDeployment .

Reply:
Proxy is removed for now likely moving to InstallEnvironment.